Do not dump duplicate configurations or Uberblocks with zdb -lu

ofaaland opened the pull request:
### Description
While looping over the labels during zdb -l, record the checksum of the config nvlist.  Dump the first configuration which can be unpacked.  For the remaining labels, dump the configuration only if it is different than config(s) previously dumped.

In the same manner, detect Uberblocks which have already been dumped, when zdb is run with -u.  The Uberblock is checked against the Uberblocks in the same slot in other labels.  

See commit message for effect of multiple -l or -u flags.

### Motivation and Context
Current behavior is to dump the configuration and entire uberblock ring for each label.  When they are not all the same, it is not obvious that one or more differ from the others.  When they are all the same, this makes the output unnecessarily long.

### How Has This Been Tested?
There are three new tests to verify the new functionality.  They create a pool with multiple vdevs, and use dd to copy label 0 from one vdev to the other.  One test checks the config output, another the uberblock output, and a third the exit code.  See  zfs-tests/tests/functional/cli_root/zdb/zdb_00{3,4,5}_pos.ksh for details

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the ZFS on Linux code style requirements.
- [x ] I have updated the documentation accordingly.
- [x ] I have read the **CONTRIBUTING** document.
- [ x] I have added tests to cover my changes.
- [x ] All new and existing tests passed.
- [x] Change has been approved by a ZFS on Linux member.

dweeezil commented:
This is a nice and reasonable feature.  How about extending it to handle the common case of ```zdb -lu```.

richardelling commented on tests/zfs-tests/tests/functional/cli_root/zdb/zdb_003_pos.ksh:
```diff
@@ -0,0 +1,90 @@
+VDEV2=$mntpnt/vdev2;
+
+log_must $MKFILE 64m $VDEV1 $VDEV2
+run_and_verify -p "$TESTPOOL" "$ZPOOL create $TESTPOOL mirror $VDEV1 $VDEV2"
```
It might be a stretch to assume xxd is always available. An alternate approach is to create a mirrored pool, as here, then create a pool with a single vdev. This will ensure the two label sets are different. Then simply dd one label from one vdev to the other. Unlike xxd, dd is always available.

ofaaland replied:
Good idea!  Thanks.

ofaaland replied:
In fact, its good enough for me to create one pool with mirror and copy a label from one vdev to another.  Much simpler.  Thanks again.

don-brady commented on cmd/zdb/zdb.c:
```diff
@@ -2360,6 +2360,42 @@ dump_nvlist_stats(nvlist_t *nvl, size_t cap)
+
+	if (nvlist_size(config, &config_size, NV_ENCODE_XDR) < 0)
+		config_size = buflen;
+
```
`nvlist_size()` returns positive errno values. As coded won't this will leave `config_size` uninitialized on errors?

ofaaland replied:
Thanks for catching that, I'll fix it.

ofaaland commented:
@dweeezil , that's a good idea.  I'll look at that.

ofaaland commented:
@behlendorf or @dinatale2 , looks like two builders are having trouble:
Amazon 2015.09 x86_64 (BUILD)
Amazon 2015.09 x86_64 SIMD (TEST)
Both haven't processed a build in about 4.5 hours and seem to be unable to recover on their own.

ofaaland commented:
Hi @dweeezil @richardelling @don-brady ,
Two builders are stuck due to some problem external to my patch, but all tests have passed on the remaining builders and I've addressed the feedback so far.  Would you mind reviewing?  Thanks!

dweeezil commented on cmd/zdb/zdb.c:
```diff
@@ -2363,6 +2338,96 @@ dump_nvlist_stats(nvlist_t *nvl, size_t cap)
+	vd.vdev_ashift = ashift;
+	vdp->vdev_top = vdp;
+
+	for (i = 0; i < VDEV_UBERBLOCK_COUNT(vdp); i++) {
```
I was originally thinking for the uberblock ring to just checksum the whole thing and either dump none or dump them all.  With this implemention, it would be cool to not print a separate line for each duplicate but instead to keep track of the first duplicate and then say that, say, [1..X] are the same as label 0.  This certainly works but it would be nice to suppress more of the output in case of duplicates. [EDIT] I'm fine with this implementation but most of the time, the whole ring is likely to be identical to that in another label.

ofaaland replied:
I initially thought of the all-or-nothing dump as well, but opted for this method to make the different one(s) stand out more.
I agree it should be simple to describe ranges that are the same and would reduce the output dramatically.  I'll do that.

ofaaland replied:
Brian and I reworked it and ended up going with a different method of handling that.  With "-lu", you see only the unique uberblocks and configs, and with each, a line like "labels = 0 1 2 3" indicating which labels that config or uberblock is found in.

So in the normal case, with "zdb -lu", you get a listing of the config and uberblocks from label 0, all indicate "labels = 0 1 2 3", and no other output.

If you give more l's or more u's you more verbose output, ultimately getting all of them listed as they are now.

ofaaland replied:
See the new commit message for sample output.

behlendorf commented on cmd/zdb/zdb.c:
```diff
@@ -2363,18 +2338,182 @@ dump_nvlist_stats(nvlist_t *nvl, size_t cap)
+	void *node, *cookie;
+	int fd;
+
+	bzero(labels, sizeof (labels[VDEV_LABELS]));
```
This should be `bzero(labels, sizeof (labels));` or something similar so we don't zero just the first label.

ofaaland replied:
Oops.  Agreed.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zdb/zdb_003_pos.ksh:
```diff
@@ -0,0 +1,63 @@
+[[ -z $DD ]] || \
+[[ -z $GREP ]] || \
+[[ -z $ZPOOL ]] && \
+	log_fail "Missing required commands"
```
This block can be dropped from here and the other new test cases.  We've done this is the past for some test cases but only for non-standard utilities.

ofaaland replied:
OK, done.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zdb/zdb_003_pos.ksh:
```diff
@@ -0,0 +1,63 @@
+log_must $DD if=/dev/${DISK[0]} of=/dev/${DISK[1]} bs=1K count=256 conv=notrunc
+
+for x in 0 1 ; do
+	config_count=$($ZDB -l /dev/${DISK[$x]} | $GREP -c features_for_read)
```
You can use `DEV_RDSKDIR` for `/dev` since Illumos expects a different path.

ofaaland replied:
Thanks.  I wondered about that but forgot to ask.

dweeezil commented:
LGTM.  I'd suggest getting this into the PR queue for openzfs as well.

behlendorf commented:
@dweeezil good idea, let's get's @ahrens thoughts on this.
